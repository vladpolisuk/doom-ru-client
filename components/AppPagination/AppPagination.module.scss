.app_pagination {
	display: flex;
	gap: 5px;

	&_item {
		display: flex;
		justify-content: center;
		align-items: center;
		height: 48px;
		min-width: 48px;
		border-radius: 50%;
		color: var(--color-2);
		font-weight: bold;
		box-sizing: border-box;
		font-size: 15px;
		background: var(--background_actives);
		border: 1px solid var(--background_actives-2);

		&[data-selected='true'] {
			background-color: var(--color-brand);
			color: var(--color);
		}
	}

	&_dots {
		display: flex;
		align-items: center;
		justify-content: center;
		height: 48px;
		min-width: 48px;
		font-size: 14px;
		color: var(--color-2);
	}
}
