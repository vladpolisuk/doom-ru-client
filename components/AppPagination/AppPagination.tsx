import clsx from 'clsx';
import { useTranslation } from 'next-i18next';
import { FC, HTMLAttributes, memo } from 'react';
import { FaChevronLeft, FaChevronRight } from 'react-icons/fa';
import usePagination, { DOTS } from '../../hooks/usePagination';
import { BaseAppComponent } from '../../types/components';
import AppButton from '../AppButton/AppButton';
import s from './AppPagination.module.scss';

export type IAppPagination = HTMLAttributes<HTMLUListElement> &
	BaseAppComponent<HTMLUListElement> & {
		/**
		 * Represents the total count of data available from the source.
		 */
		totalPages: number;
		/**
		 * Represents the maximum data that is visible in a single page.
		 */
		pageSize: number;
		/**
		 * Represents the current active page. Defaults to 1.
		 */
		currentPage?: number;
		/**
		 * Represents the min number of page buttons to be shown on each side of the current page button. Defaults to 1.
		 */
		siblingCount?: number;
		/**
		 * Callback function invoked with the updated page value when the page is changed.
		 * @param page number
		 */
		onPageChange: (page: number) => void;
	};

/**
 * The common pagination component in the application
 * @type `FC<AppPagination>`
 * @memo `true`
 * @return `html:ul`
 */
const AppPagination: FC<IAppPagination> = memo(
	({
		title,
		children,
		className = '',
		totalPages,
		pageSize,
		currentPage = 1,
		siblingCount = 1,
		onlyARIA = false,
		resetStyles = false,
		onPageChange,
		...props
	}) => {
		const t = useTranslation('search').t;

		const paginationRange = usePagination({
			currentPage,
			totalPages,
			siblingCount,
			pageSize
		});

		if (!paginationRange) return null;
		if (currentPage === 0 || paginationRange.length < 2) return null;

		const onNext = () => onPageChange(currentPage + 1);
		const onPrevious = () => onPageChange(currentPage - 1);

		const prevLabel = t('search_result_btn.prev.label');
		const nextLabel = t('search_result_btn.next.label');
		const pageLabel = (page: number) => t('search_result_btn.page.label', { page });
		const lastPage = paginationRange[paginationRange.length - 1];
		const titleAttr = onlyARIA ? '' : title;
		const paginationStyles = clsx(className, s.app_pagination, 'unlisted');
		const itemStyles = clsx(s.app_pagination_item, 'active--scale');
		const leftBtnStyles = clsx(s.app_pagination_item, currentPage !== 1 && 'active--scale');
		const rightBtnStyles = clsx(s.app_pagination_item, currentPage !== lastPage && 'active--scale');

		return (
			<ul
				className={paginationStyles}
				aria-label={title}
				title={titleAttr}
				data-testid='app-pagination'
				{...props}>
				<li>
					<AppButton
						onlyARIA
						color='transparent'
						title={prevLabel}
						disabled={currentPage === 1}
						className={leftBtnStyles}
						onClick={onPrevious}>
						<FaChevronLeft />
					</AppButton>
				</li>

				{paginationRange.map((pageNumber, i) =>
					pageNumber === DOTS ? (
						<li
							key={`dots-${i}`}
							className={s.app_pagination_dots}>
							{DOTS}
						</li>
					) : (
						<li key={pageNumber}>
							<AppButton
								onlyARIA
								color='transparent'
								className={itemStyles}
								title={pageLabel(pageNumber as number)}
								data-selected={pageNumber === currentPage}
								onClick={() => onPageChange(pageNumber as number)}>
								{pageNumber}
							</AppButton>
						</li>
					)
				)}

				<li>
					<AppButton
						onlyARIA
						color='transparent'
						title={nextLabel}
						disabled={currentPage === lastPage}
						className={rightBtnStyles}
						onClick={onNext}>
						<FaChevronRight />
					</AppButton>
				</li>
			</ul>
		);
	}
);

AppPagination.displayName = 'AppPagination';
export default AppPagination;
