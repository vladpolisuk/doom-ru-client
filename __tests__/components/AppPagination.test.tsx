import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import AppPagination from '../../components/AppPagination/AppPagination';

describe('AppPagination', () => {
	it('should be rendered', () => {
		const onPageChange = jest.fn();
		const wrapper = render(
			<AppPagination
				totalPages={100}
				currentPage={10}
				pageSize={5}
				onPageChange={onPageChange}
			/>
		);
		const pagination = screen.getByTestId('app-pagination');
		expect(pagination).toBeInTheDocument();
		expect(wrapper).toMatchSnapshot();
	});
});
