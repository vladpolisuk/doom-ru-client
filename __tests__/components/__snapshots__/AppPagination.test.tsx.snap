// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppPagination should be rendered 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="app_pagination unlisted"
        data-testid="app-pagination"
      >
        <li>
          <button
            aria-label="search_result_btn.prev.label"
            class="app_button app_button--transparent app_pagination_item active--scale"
            data-testid="app-button"
            title=""
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-label="search_result_btn.page.label"
            class="app_button app_button--transparent app_pagination_item active--scale"
            data-selected="false"
            data-testid="app-button"
            title=""
          >
            1
          </button>
        </li>
        <li
          class="app_pagination_dots"
        >
          • • •
        </li>
        <li>
          <button
            aria-label="search_result_btn.page.label"
            class="app_button app_button--transparent app_pagination_item active--scale"
            data-selected="false"
            data-testid="app-button"
            title=""
          >
            9
          </button>
        </li>
        <li>
          <button
            aria-label="search_result_btn.page.label"
            class="app_button app_button--transparent app_pagination_item active--scale"
            data-selected="true"
            data-testid="app-button"
            title=""
          >
            10
          </button>
        </li>
        <li>
          <button
            aria-label="search_result_btn.page.label"
            class="app_button app_button--transparent app_pagination_item active--scale"
            data-selected="false"
            data-testid="app-button"
            title=""
          >
            11
          </button>
        </li>
        <li
          class="app_pagination_dots"
        >
          • • •
        </li>
        <li>
          <button
            aria-label="search_result_btn.page.label"
            class="app_button app_button--transparent app_pagination_item active--scale"
            data-selected="false"
            data-testid="app-button"
            title=""
          >
            20
          </button>
        </li>
        <li>
          <button
            aria-label="search_result_btn.next.label"
            class="app_button app_button--transparent app_pagination_item active--scale"
            data-testid="app-button"
            title=""
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="app_pagination unlisted"
      data-testid="app-pagination"
    >
      <li>
        <button
          aria-label="search_result_btn.prev.label"
          class="app_button app_button--transparent app_pagination_item active--scale"
          data-testid="app-button"
          title=""
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
            />
          </svg>
        </button>
      </li>
      <li>
        <button
          aria-label="search_result_btn.page.label"
          class="app_button app_button--transparent app_pagination_item active--scale"
          data-selected="false"
          data-testid="app-button"
          title=""
        >
          1
        </button>
      </li>
      <li
        class="app_pagination_dots"
      >
        • • •
      </li>
      <li>
        <button
          aria-label="search_result_btn.page.label"
          class="app_button app_button--transparent app_pagination_item active--scale"
          data-selected="false"
          data-testid="app-button"
          title=""
        >
          9
        </button>
      </li>
      <li>
        <button
          aria-label="search_result_btn.page.label"
          class="app_button app_button--transparent app_pagination_item active--scale"
          data-selected="true"
          data-testid="app-button"
          title=""
        >
          10
        </button>
      </li>
      <li>
        <button
          aria-label="search_result_btn.page.label"
          class="app_button app_button--transparent app_pagination_item active--scale"
          data-selected="false"
          data-testid="app-button"
          title=""
        >
          11
        </button>
      </li>
      <li
        class="app_pagination_dots"
      >
        • • •
      </li>
      <li>
        <button
          aria-label="search_result_btn.page.label"
          class="app_button app_button--transparent app_pagination_item active--scale"
          data-selected="false"
          data-testid="app-button"
          title=""
        >
          20
        </button>
      </li>
      <li>
        <button
          aria-label="search_result_btn.next.label"
          class="app_button app_button--transparent app_pagination_item active--scale"
          data-testid="app-button"
          title=""
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
            />
          </svg>
        </button>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
